### DISCUSSION TITLE
How do you handle random connection failures?

### DISCUSSION THREAD

**Discussion by laurent22** [community/community#62294] [Actions] [answered]
### Select Topic Area

Question

### Body

As our CI (based on GitHub Actions) is getting bigger, we often have random connection failures, for example some package or archive somewhere cannot be downloaded. It can be resolved by re-running the failed jobs but this is annoying and it has to be done quite frequently.

Going through the scripts and code and add request-repeating logic everywhere is not an option, so I'm wondering how do you deal with this in general? Are there any general recommendations that can be followed to make GitHub CI more reliable?

**Comment by h7x4** [comment, marked as answer]
It depends on the intention behind the connections.

For downloading packages and 3rd party dependencies (as seems to be your case), caching is probably the best answer. There is a documentation section dedicated to this topic:

https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows

But if the connection errors are caused by integration tests that uses network connections, it's more complicated. You could try to reduce the chance of failure by doing the integration tests on the same machine and using sockets on a loopback device (i.e `localhost`). Keep in mind that this might reduce the similarity to the real world of the tests, and it might make the tests catch fewer errors. I think this topic is quite broad and complex, so if this is your case I think you should specify it.
